inference-engine/src/transformations/src/transformations/common_optimizations/nop_elimination.cpp

```diff
@@ -503,6 +504,123 @@ pass::EliminateTranspose::EliminateTranspose() {
     this->register_matcher(m, callback);
 }
 
+NGRAPH_RTTI_DEFINITION(pass::RemoveConcatZeroDimInput, "RemoveConcatZeroDimInput", 0);
+
+pass::RemoveConcatZeroDimInput::RemoveConcatZeroDimInput() {
+    MATCHER_SCOPE(RemoveConcatZeroDimInput);
+    auto concat_pattern = pattern::wrap_type<opset8::Concat>();
+    ngraph::matcher_pass_callback callback = [=](pattern::Matcher& m) {
+        auto concat = m.get_match_root();
+        auto fixed_concat_inputs = concat->input_values();
+        fixed_concat_inputs.erase(std::remove_if(fixed_concat_inputs.begin(), fixed_concat_inputs.end(),
+            [](const Output<Node>& input) {
+                const auto& in_shape = input.get_partial_shape();
+                if (in_shape.rank().is_static()) {
+                    return std::any_of(std::begin(in_shape), std::end(in_shape), [](const ov::Dimension& dim) {
+                        if (dim.is_static() && dim.get_length() == 0) {
+                            return true;
+                        }
+                        return false;
+                    });}
+                    return false;
+                }), fixed_concat_inputs.end());
+
+        bool inputs_removed = concat->get_input_size() > fixed_concat_inputs.size();
+        if (inputs_removed) {
+            concat->set_arguments(fixed_concat_inputs);
+        }
+        return false;
+    };
+    auto m = std::make_shared<ngraph::pattern::Matcher>(concat_pattern, matcher_name);
+    this->register_matcher(m, callback);
+}
+
+NGRAPH_RTTI_DEFINITION(ngraph::pass::RemoveMultiSubGraphOpDanglingParams, "RemoveMultiSubGraphOpDanglingParams", 0);
+
+ngraph::pass::RemoveMultiSubGraphOpDanglingParams::RemoveMultiSubGraphOpDanglingParams() {
+    MATCHER_SCOPE(RemoveMultiSubGraphOpDanglingParams);
+    auto multi_subgraph_op_pattern = pattern::wrap_type<
```

Comment:
I guess `op::util::MultiSubGraphOp` is enough here. Loop, TI and If are inherited from this class.

Reply:
Unfortunately, using here only base class `op::util::MultiSubGraphOp` doesn't work (pattern is not detected)

Reply:
Looks like for Loop and TI operations RTTI is not defined properly.
https://github.com/openvinotoolkit/openvino/blob/master/ngraph/core/include/openvino/op/tensor_iterator.hpp#L19
Needs to be changed to `OPENVINO_OP("TensorIterator", "opset1", op::util::SubGraphOp, op::util::MultiSubGraphOp);`

Reply:
good catch, done